### PR TITLE
[OCTREE] Fix pcl_octree_viewer

### PR DESCRIPTION
**Pull request by frozar**
Fixes issue #1840 

To avoid the display of spurious point, check the detph of an octree node before to print it.
Be able to read PLY file.
Fix the display of the octree cube.

### CODE REVIEW

**Comment by taketwo** [comment]
Hi, thanks for contributing. If I understand correctly, you propose three different changes here. It is okay to keep them in a single pull request, but would be better to split into three commits with appropriate messages. Further, I'd ask you to follow the [PCL style guide](http://www.pointclouds.org/documentation/advanced/pcl_style_guide.php). (In particular, spacing in function calls and positioning of curly brackets.) Please update this pull request.

**Comment by frozar** [comment]
I split my contribution in 3 commits as you advise. I try to follow the PCL coding style but if there still is some coding style issue, please put some comment on the concern lines and I'll fix them.

I must say the code I have written to fix the spurious points display is pretty odd: if the iterator over an octree was correct, I should not have to check the depth of the iterator (`tree_it.getCurrentOctreeDepth () != depth`). Don't you think that there is an issue about the octree iterator?

**Comment by taketwo** [comment]
From what I see, the argument for `begin()` function is called `max_depth_arg`, so this is expected that you get nodes with depth less or equal to what is passed to the function.

**Comment by taketwo** [comment]
Regarding the file loading commit. I've just remembered that we have a relatively new [`load`](http://docs.pointclouds.org/trunk/auto__io_8hpp_source.html#l00055) function that automatically deals with all supported filetypes.

**Comment by frozar** [comment]
I agree with you for the `max_depth_arg`. I think the first person who wrote `octree_viewer.cpp` didn't notice that. Don't you think that it would be valuable to provide an other iterator that does the work? (I prepare a merge request in this way, but I don't if this feature will be merged (I send a mail on the PCL-users mailing list, but no answer yet...)).

I update the code with the `load` function.

**Comment by taketwo** [comment]
> Don't you think that it would be valuable to provide an other iterator that does the work?

Honestly, I don't have a strong opinion here. The octree module has been around for many years now and nobody submitted either a feature request or an implementation of such an iterator. So it seems fine without. That said, if you find multiple places in existing code in PCL that can be rewritten in a more clear and concise way by using the proposed iterator, I would take it as a strong evidence that the iterator is valuable and will be happy to merge it. Of course, other opinions are welcome @SergioRAgostinho @jspricke 

> I send a mail on the PCL-users mailing list, but no answer yet...

This issue tracker is the right place to propose and discuss new functionality before actually crafting a pull request.

File: tools/octree_viewer.cpp [hunk 1]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
-  }
 
   /* \brief display octree cubes via vtk-functions
    *
    */
   void showCubes(double voxelSideLen)
   {
-    //get the renderer of the visualizer object
-    vtkRenderer *renderer = viz.getRenderWindow()->GetRenderers()->GetFirstRenderer();
+    vtkSmartPointer<vtkAppendPolyData> appendFilter = vtkSmartPointer<vtkAppendPolyData>::New();
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Need space before brackets, i.e. `New ();`

File: tools/octree_viewer.cpp [hunk 2]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
 
-    vtkSmartPointer<vtkAppendPolyData> treeWireframe = vtkSmartPointer<vtkAppendPolyData>::New();
-    size_t i;
+    // Create every cubes to be displayed
     double s = voxelSideLen / 2.0;
-    for (i = 0; i < displayCloud->points.size(); i++)
+    for (size_t i = 0; i < displayCloud->points.size(); i++)
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Need space before brackets, i.e. `points.size ();`

File: tools/octree_viewer.cpp [hunk 3]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
     {
-
       double x = displayCloud->points[i].x;
       double y = displayCloud->points[i].y;
       double z = displayCloud->points[i].z;
 
+      vtkSmartPointer<vtkCubeSource> wk_cubeSource = vtkSmartPointer<vtkCubeSource>::New();
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Space

File: tools/octree_viewer.cpp [hunk 4]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
+
+      wk_cubeSource->SetBounds(x - s, x + s, y - s, y + s, z - s, z + s);
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Space here and in all line.s below.

File: tools/octree_viewer.cpp [hunk 5]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
+
+    multiMapper->SetInputConnection(cleanFilter->GetOutputPort());
+
+    vtkSmartPointer<vtkActor> multiActor = vtkSmartPointer<vtkActor>::New();
+
+    multiActor->SetMapper(multiMapper);
 
-    treeActor->GetProperty()->SetColor(1.0, 1.0, 1.0);
-    treeActor->GetProperty()->SetLineWidth(2);
+    multiActor->GetProperty( )->SetColor(0.0, 1.0, 1.0);
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: No spaces inside brackets

File: tools/octree_viewer.cpp [hunk 6]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
+    multiActor->GetProperty( )->SetAmbient(1.0);
+    multiActor->GetProperty( )->SetLineWidth(2);
     if(wireframe)
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Space after `if`

**Comment by frozar** [comment]
Maybe you should check the coding style again: I never wrote code like that... ^^'

File: tools/octree_viewer.cpp [hunk 7]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
       double z = displayCloud->points[i].z;
 
+      vtkSmartPointer<vtkCubeSource> wk_cubeSource = vtkSmartPointer<vtkCubeSource>::New ();
+
+      wk_cubeSource->SetBounds (x - s, x + s, y - s, y + s, z - s, z + s);
+      wk_cubeSource->Update ();
+
 #if VTK_MAJOR_VERSION < 6
-      treeWireframe->AddInput(GetCuboid(x - s, x + s, y - s, y + s, z - s, z + s));
+      appendFilter->AddInput(wk_cubeSource->GetOutput ());
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Missing space after `AddInput`.

File: tools/octree_viewer.cpp [hunk 8]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
 #else
-      treeWireframe->AddInputData (GetCuboid (x - s, x + s, y - s, y + s, z - s, z + s));
+      appendFilter->AddInputData(wk_cubeSource->GetOutput ());
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Same

File: tools/octree_viewer.cpp [hunk 9]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
+    vtkSmartPointer<vtkCleanPolyData> cleanFilter = vtkSmartPointer<vtkCleanPolyData>::New ();
 
-    vtkSmartPointer<vtkDataSetMapper> mapper = vtkSmartPointer<vtkDataSetMapper>::New();
-#if VTK_MAJOR_VERSION < 6
-    mapper->SetInput(treeWireframe->GetOutput());
-#else
-    mapper->SetInputData (treeWireframe->GetOutput ());
-#endif
-    treeActor->SetMapper(mapper);
+    cleanFilter->SetInputConnection(appendFilter->GetOutputPort ());
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Missing space after `SetInputConnection`

File: tools/octree_viewer.cpp [hunk 10]
```diff
@@ -265,62 +265,68 @@ class OctreeViewer
...
+    multiMapper->SetInputConnection (cleanFilter->GetOutputPort ());
+
+    vtkSmartPointer<vtkActor> multiActor = vtkSmartPointer<vtkActor>::New ();
+
+    multiActor->SetMapper (multiMapper);
 
-    treeActor->GetProperty()->SetColor(1.0, 1.0, 1.0);
-    treeActor->GetProperty()->SetLineWidth(2);
-    if(wireframe)
+    multiActor->GetProperty ()->SetColor (0.0, 1.0, 1.0);
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Any particular reason to change the color from white to cyan?

File: tools/octree_viewer.cpp [hunk 11]
```diff
@@ -339,6 +345,9 @@ class OctreeViewer
 
     for (tree_it = octree.begin(depth); tree_it!=tree_it_end; ++tree_it)
     {
+      if (tree_it.getCurrentOctreeDepth () != depth)
+        continue ;
```
Comment (taketwo) [inline tools/octree_viewer.cpp]: Extra space before semicolon.

File: tools/octree_viewer.cpp (hunk 10 above)
Comment (frozar) [reply tools/octree_viewer.cpp]: Not at all. In fact, I was just trying to find the color used to produce the image [there](http://docs.pointclouds.org/trunk/group__octree.html), but unsuccessfully...

Comment (taketwo) [reply tools/octree_viewer.cpp]: Ok, then let's leave it white as it was before.

**Comment by taketwo** [comment]
So besides from color everything looks good now. Once that is fixed, please regroup all related changes so that we have 3 topic commits, as planned, and then we are good to go.

**Comment by frozar** [comment]
I rewrite the history, now it should be fine :ok_hand:

File: tools/octree_viewer.cpp (hunk 10 above)
Comment (frozar) [reply tools/octree_viewer.cpp]: Oh, I didn't see this message...

**Comment by taketwo** [comment]
Thanks for the effort. From my side everything is fine, but we'll need to wait a bit until one other maintainer takes a look and pushes the merge button.

**Comment by frozar** [comment]
Nice :smile:

**Comment by SergioRAgostinho** [comment]
> Of course, other opinions are welcome @SergioRAgostinho

+1 for the iterator

**Pull request merged**